File: output/playwright/nested-wires/capture-m45.py
"""Only isolated, nonpersistent headless Playwright; existing Vite is never restarted."""
import json
import subprocess
from pathlib import Path
ROOT = Path(__file__).resolve().parents[3]
OUT = ROOT / 'output/playwright/nested-wires'
CLI = Path.home() / '.codex/skills/playwright/scripts/playwright_cli.sh'
BASELINE = ROOT / '.local/m45/baseline-scene.json'
BASELINE.parent.mkdir(parents=True, exist_ok=True)
BASELINE.write_bytes(subprocess.check_output(['git', 'show', '0e5f21e:output/playwright/nested-wires/scene.json'], cwd=ROOT))
def command(*args):
    result = subprocess.run([str(CLI), '-s=m45-capture', *args], cwd=ROOT, text=True, capture_output=True, check=True)
    if '### Error' in result.stdout: raise RuntimeError(result.stdout)
    return result.stdout
try:
    opened = command('open', 'http://127.0.0.1:5188/roads-prototype.html?nested')
    command('snapshot')
    source = (OUT / 'capture-m45.mjs').read_text().replace('export async function', 'async function')
    result = command('run-code', 'async () => {\n' + source + '\nreturn capture(page,' + json.dumps('/@fs/' + str(BASELINE)) + ');\n}')
    report = json.loads(result.split('### Result\n', 1)[1].split('\n###', 1)[0])
    (OUT / 'm45-browser.json').write_text(json.dumps(report, indent=2)+'\n')
    images = [f'http://127.0.0.1:5188/@fs/{OUT}/m45-junction-{side}.png' for side in ['before','after']]
    html = '<html><body style="margin:0;background:#fafafa;font:18px system-ui;display:flex;gap:16px;padding:20px">' + ''.join(f'<div><p>M4{".5" if i else ""} · J21 · {2 if i else 6} crossings</p><img width="448" src="{src}"></div>' for i,src in enumerate(images)) + '</body></html>'
    command('run-code', 'async () => { await page.setViewportSize({width:952,height:550}); await page.setContent(' + json.dumps(html) + '); await page.locator("img").evaluateAll(xs => Promise.all(xs.map(x => x.decode()))); await page.screenshot({path:' + json.dumps(str(OUT / 'm45-junction-before-after.png')) + '}); }')
    assert report['medianMilliseconds'] <= 300, report
    assert report['nodes'] == 24 and report['wires'] == 26 and report['layoutRecalcCount'] == 1 and report['roadsOff']
    print(f"PASS five loads={[l['roads:navigation-to-ready'] for l in report['loads']]}; median={report['medianMilliseconds']:.3f} ms <=300")
    print('PASS headless captures: immutable M4 before / current M4.5 after, overview, roads-off; 24 nodes / 26 wires; layout count 1')
finally:
    command('close')

File: output/playwright/nested-wires/capture-m45.mjs
/** Headless capture of current output and the immutable M4 scene in the same unchanged renderer. */
const directory = 'output/playwright/nested-wires/';
const url = 'http://127.0.0.1:5188/roads-prototype.html?nested';
async function ready(page) {
  await page.goto(url);
  await page.waitForFunction(() => performance.getEntriesByName('roads:navigation-to-ready').length > 0);
}
async function junctionCapture(page, name) {
  const junction = page.getByRole('button', { name: /^Inspect J21 / });
  const before = await junction.boundingBox();
  await page.mouse.move(before.x + before.width / 2, before.y + before.height / 2);
  await page.mouse.wheel(0, -1600);
  await page.waitForTimeout(400);
  const box = await junction.boundingBox();
  await page.screenshot({ path: `${directory}${name}.png`, clip: { x: box.x - 40, y: box.y - 40, width: box.width + 80, height: box.height + 80 } });
  return box;
}
export async function capture(page, baselineUrl) {
  await page.setViewportSize({ width: 1920, height: 1440 });
  const loads = [];
  for (let i = 0; i < 5; i += 1) {
    await ready(page);
    loads.push(await page.evaluate(() => Object.fromEntries(performance.getEntriesByType('measure').filter((e) => e.name.startsWith('roads:')).map((e) => [e.name, e.duration]))));
  }
  await page.waitForTimeout(250);
  await page.screenshot({ path: `${directory}m45-overview.png` });
  const checkbox = page.getByRole('checkbox', { name: 'Show roads' });
  await checkbox.uncheck();
  await page.screenshot({ path: `${directory}m45-roads-off.png` });
  const roadsOff = !(await checkbox.isChecked());
  await checkbox.check();
  const after = await junctionCapture(page, 'm45-junction-after');
  const report = { loads, medianMilliseconds: loads.map((l) => l['roads:navigation-to-ready']).toSorted((a,b) => a-b)[2], roadsOff,
    nodes: await page.locator('[data-node-id]').count(), wires: await page.locator('[data-wire-id]').count(),
    layoutRecalcCount: await page.evaluate(() => window.__layoutRecalcCount), after };
  let intercepted = 0;
  await page.route('**/cli/roads-prototype.ts*', async (route) => {
    intercepted += 1;
    const response = await route.fetch();
    const source = await response.text();
    const expression = 'const scene = build({ measure });';
    if (!source.includes(expression)) throw new Error('Baseline harness cannot locate builder expression');
    const body = source.replace(expression, `const scene = await (await fetch(${JSON.stringify(baselineUrl)})).json();`);
    await route.fulfill({ response, body });
  });
  await ready(page);
  await page.waitForTimeout(250);

  report.baselineVerified = await page.evaluate(async (url) => { const baseline = await (await fetch(url)).json(); return baseline.wiring.value.every((wire) => { const points = [wire.segments[0].from, ...wire.segments.map((s) => s.to)].map((p) => `${p.x},${p.y}`).join(' '); return document.querySelector(`[data-wire-hit="${wire.id}"]`).getAttribute('points') === points; }); }, baselineUrl);
  verifyBaseline(report, intercepted);
  report.before = await junctionCapture(page, 'm45-junction-before');
  await page.unroute('**/cli/roads-prototype.ts*');
  return report;
}

function verifyBaseline(report, intercepted) {
  if (intercepted !== 1) throw new Error(`Baseline app intercept count ${intercepted}`);
  if (!report.baselineVerified) throw new Error('Rendered baseline wire paths differ from immutable M4 scene');
}
